### DISCUSSION TITLE
Is it possible to preview and autoformat qhelp files locally?

### DISCUSSION THREAD

**Discussion by p-** [github/securitylab#62] [General] [answered]
Hi, I thought I'd ask this here.

Pull requests made against the [ql](https://github.com/Semmle/ql/) repository are tested on different internal Jenkins jobs. One of them is QHelp-Preview.
Is it possible to preview the QHelp files locally?

Also there seems to be a QHelp autoformat feature. Where is it located?
(The VS Code extension doesn't seem to have a formatter for QHelp files)

**Comment by p0** [comment, marked as answer]
It is currently not possible to render QHelp files with the public `codeql` CLI, unfortunately; we're planning to add the necessary functionality, but so far it hasn't been the top of the priority list. We'll hopefully expose it through the CLI and the VS Code extension in the coming weeks.

We don't have autoformatting for QHelp internally, but it's just XML, so perhaps an XML formatter might do a good enough job?

**Reply by p-** [reply]
Hi @p0 
Thanks for answering so fast!

> hasn't been the top of the priority list

I agree it's probably not that an important feature.

> but it's just XML, so perhaps an XML formatter might do a good enough job

Good point! I ended up not using an XML formatter as they typically don't do that left aligned style of QHelp files. Anyways, the XML formatter helped finding my embarassing mistake ;)
